plugin/gcp_account_resources.go

```diff
@@ -131,6 +150,146 @@ func (b *backend) createServiceAccount(ctx context.Context, req *logical.Request
+			w := fmt.Sprintf("unable to delete key under service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+			warnings = append(warnings, w)
+		}
+	}
+
+	if merr := b.removeBindings(ctx, req, boundResources.accountId.EmailOrId, boundResources.bindings); merr != nil {
+		for _, err := range merr.Errors {
+			w := fmt.Sprintf("unable to delete IAM policy bindings for service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.EmailOrId, err)
+			warnings = append(warnings, w)
+		}
+	}
+
+	if removeServiceAccount {
+		if err := b.deleteServiceAccount(ctx, iamAdmin, boundResources.accountId); err != nil {
+			w := fmt.Sprintf("unable to delete service account %q (WAL entry to clean-up later has been added): %v", boundResources.accountId.ResourceName(), err)
+			warnings = append(warnings, w)
+		}
+	}
+
+	// If resources were deleted, we don't need the WAL rollbacks we created for these resources.
+	if len(warnings) == 0 {
+		b.tryDeleteWALs(ctx, req.Storage, walIds...)
+	}
+
+	return nil
+}
+
+func (b *backend) deleteTokenGenKey(ctx context.Context, iamAdmin *iam.Service, tgen *TokenGenerator) error {
+	if tgen == nil || tgen.KeyName == "" {
+		return nil
+	}
+
+	_, err := iamAdmin.Projects.ServiceAccounts.Keys.Delete(tgen.KeyName).Do()
+	if err != nil && !isGoogleAccountKeyNotFoundErr(err) {
+		return errwrap.Wrapf("unable to delete service account key: {{err}}", err)
+	}
+	return nil
+}
+
+func (b *backend) removeBindings(ctx context.Context, req *logical.Request, email string, bindings ResourceBindings) (allErr *multierror.Error) {
+	httpC, err := b.HTTPClient(req.Storage)
+	if err != nil {
+		return &multierror.Error{Errors: []error{err}}
+	}
+
+	apiHandle := iamutil.GetApiHandle(httpC, useragent.String())
+
+	for resName, roles := range bindings {
+		resource, err := b.resources.Parse(resName)
+		if err != nil {
+			allErr = multierror.Append(allErr, errwrap.Wrapf(fmt.Sprintf("unable to delete role binding for resource '%s': {{err}}", resName), err))
+			continue
+		}
+
+		p, err := resource.GetIamPolicy(ctx, apiHandle)
+		if err != nil {
+			allErr = multierror.Append(allErr, errwrap.Wrapf(fmt.Sprintf("unable to delete role binding for resource '%s': {{err}}", resName), err))
+			continue
+		}
+
+		changed, newP := p.RemoveBindings(&iamutil.PolicyDelta{
+			Email: email,
+			Roles: roles,
+		})
+		if !changed {
+			continue
+		}
+		if _, err = resource.SetIamPolicy(ctx, apiHandle, newP); err != nil {
+			allErr = multierror.Append(allErr, errwrap.Wrapf(fmt.Sprintf("unable to delete role binding for resource '%s': {{err}}", resName), err))
+			continue
+		}
+	}
+	return
+}
+
+func (b *backend) deleteServiceAccount(ctx context.Context, iamAdmin *iam.Service, account gcputil.ServiceAccountId) error {
+	if account.EmailOrId == "" {
+		return nil
+	}
+
+	_, err := iamAdmin.Projects.ServiceAccounts.Delete(account.ResourceName()).Do()
```

Comment:
I realize this is existing code that's been moved to a new file, but I think we should be passing the context through to external API calls. Looks like this can be done with `.Context(ctx).Do()`.

Here are some locations (there may be more) where the context is unused:
- `deleteServiceAccount()`
- `createNewTokenGen()`
- `createServiceAccount()`
- `deleteTokenGenKey()`
- `secretKeyRevoke()`

Reply:
Sure. I've added them.